hal/armv8m/mcx/_init.S

```diff
@@ -0,0 +1,359 @@
+.word _start
+
+.word _exceptions_dispatch /* NMI */
+.word _exceptions_dispatch /* HardFault */
+.word _exceptions_dispatch /* MemMgtFault */
+.word _exceptions_dispatch /* BusFault */
+.word _exceptions_dispatch /* UsageFault */
+.word _exceptions_dispatch /* SecureFault */
+.word 0
+.word 0
+.word 0
+.word _syscall_dispatch    /* SVC */
+.word _exceptions_dispatch /* Debug */
+.word 0
+.word _interrupts_dispatch /* PendSV */
+.word _interrupts_dispatch /* Systick */
+
+.rept 156                  /* Max number of ext interrupts - last peripheral id + 1 */
+.word _interrupts_dispatch
+.endr
+.size _init_vectors, .-_init_vectors
+
+
+.thumb
+.thumb_func
+
+.globl _start
+.type _start, %function
+
+_start:
+	cpsid if
+	isb
+
+	/* Point to syspage */
+	ldr r8, =syspage
+	str r9, [r8]
+
+	/* Init vector table pointer */
+	ldr r0, =0xe000ed08u
+	ldr r1, =_init_vectors
+	str r1, [r0]
+	isb
+	dmb
+
+	/* Init MSP to a first value in _init_vectors (.bss end + 1024 + 256) */
+	ldr r0, [r1]
+	bic r0, 7
+	msr msp, r0
+	isb
+
+	bl _mcxn94x_init
+
+	b main
+.size _start, .-_start
+.ltorg
+
+.globl _syscall_dispatch
+.type _syscall_dispatch, %function
+.type _syscallend, %function
+
+_syscall_dispatch:
+	mov r0, #0
+	msr control, r0
+	isb
+
+	mov r0, sp
+
+	/* Prepare context on kernel stack */
+	sub sp, sp, #((14 * 4) + (8 * 4))
+	str sp, [sp]
+
+	add r1, sp, #12
+	mov r12, #0
+	stmia r1!, {r4-r12}
+	str r0, [r1], #8 /* msp */
+	mov r10, r1 /* addr to store hw-saved ctx */
+
+	mrs r0, psp
+	str r0, [sp, #8]
+
+	/* Load hardware saved registers from user stack */
+	ldmia r0!, {r1-r8}
+
+	/* Fix PC LSB not being set */
+	orr r7, r7, #1
+
+	/* Store hardware saved registers on kernel stack */
+	add r10, r10, #(5 * 4)
+	stm r10, {r6-r8} /* lr, pc, psr */
+
+	/* Copy arguments back to the user stack */
+	stmdb r0!, {r1-r4}
+	mov r1, r0
+	ldrb r0, [r7, #-3]
+
+	/* Prepare pseudo context */
+	mov r7, #0x01000000
+	ldr r6, =syscalls_dispatch
+	ldr r5, =_syscallend
+	stmdb sp!, {r0-r7} /* PSR, PC, LR, R12, R3, R2, R1, R0 */
+
+	/* Exit handler mode to kernel thread mode */
+	ldr lr,=RET_THREAD_MSP
+	bx lr
+
+_syscallend:
+	/* return value in r0 */
+	ldr lr, [sp, #8] /* psp */
+	add lr, lr, #(8 * 4)
+	msr psp, lr
+	isb
+
+	add r10, sp, #((14 * 4) + (5 * 4))
+	ldm r10, {r2,r3} /* lr, pc */
+
+	add lr, sp, #12
+	ldmia lr!, {r4-r11}
+	add lr, lr, #(11 * 4)
+	mov sp, lr
+
+	/* Switch stacks */
+	mov r1, #3
+	msr control, r1
+	isb
+
+	mov lr, r2
+	mov pc, r3
+.size _syscall_dispatch, .-_syscall_dispatch
+.ltorg
+
+.globl _exceptions_dispatch
+.type _exceptions_dispatch, %function
+
+_exceptions_dispatch:
+	cpsid if
+	isb
+
+	mrs r0, psp
+	stmdb sp!, {r0, r4-r11, lr}
+
+	mrs r0, ipsr
+	mov r1, sp
+
+	b exceptions_dispatch
+.size _exceptions_dispatch, .-_exceptions_dispatch
+.ltorg
+
+.globl _interrupts_dispatch
+.type _interrupts_dispatch, %function
+_interrupts_dispatch:
+	mov r0, #0
+	msr control, r0
+	isb
+
+	mov r0, sp
+	and r12, lr, #(1 << 2)
+	cmp r12, #0
```

Comment:
Simpler
```suggestion
	ands r12, lr, #(1 << 2)
```
If the value of r12 is not used then
```suggestion
	tst  lr, #(1 << 2)
```

Reply:
Thanks, fixed